pkg/unikernel/service/sandbox.go

```diff
@@ -0,0 +1,224 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/frakti/pkg/unikernel/metadata"
+	"k8s.io/frakti/pkg/unikernel/metadata/store"
+	"k8s.io/frakti/pkg/util/uuid"
+
+	kubeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
+)
+
+// RunPodSandbox creates and starts a pod-level sandbox.
+func (u *UnikernelRuntime) RunPodSandbox(config *kubeapi.PodSandboxConfig) (string, error) {
+	var err error
+	// Genrate sandbox ID and name
+	podID := uuid.NewUUID()
+	podName := makeSandboxName(podID, config.GetMetadata())
+	// Reserve sandbox name
+	if err = u.sandboxNameIndex.Reserve(podName, podID); err != nil {
+		return "", fmt.Errorf("failed to reserve sandbox name %q: %v", podName, err)
+	}
+	defer func() {
+		if err != nil {
+			u.sandboxNameIndex.ReleaseByName(podName)
+		}
+	}()
+	// Reserve sandbox ID
+	if err = u.sandboxIDIndex.Add(podID); err != nil {
+		return "", fmt.Errorf("failed to reserve sandbox ID %q: %v", podID, err)
+	}
+	defer func() {
+		if err != nil {
+			u.sandboxIDIndex.Delete(podID)
+		}
+	}()
+
+	// TODO(Crazykev): Get cpu/mem from cgroup
+	vmMeta := &metadata.VMMetadata{
+		CPUNum: u.defaultCPU,
+		Memory: u.defaultMem,
+	}
+
+	// Create sandbox metadata.
+	meta := metadata.SandboxMetadata{
+		ID:       podID,
+		Name:     podName,
+		Config:   config,
+		VMConfig: vmMeta,
+	}
+
+	// TODO(Crazykev): Create ns and cni config
+
+	// Add sandbox into sandbox metadata store.
+	meta.CreatedAt = time.Now().UnixNano()
+	meta.State = kubeapi.PodSandboxState_SANDBOX_READY
+	if err = u.sandboxStore.Create(meta); err != nil {
+		return "", fmt.Errorf("failed to add sandbox metadata %+v into store: %v", meta, err)
+	}
+
+	return podID, nil
+}
+
+// StopPodSandbox stops the sandbox. If there are any running containers in the
+// sandbox, they should be force terminated.
+func (u *UnikernelRuntime) StopPodSandbox(podSandboxID string) error {
+	sandbox, err := u.sandboxStore.Get(podSandboxID)
+	if err != nil {
+		return fmt.Errorf("failed to find sandbox(%q): %v", podSandboxID, err)
+	}
+	// Stop relate VM
+	if err = u.vmTool.StopVM(sandbox.ID, 0); err != nil {
+		return fmt.Errorf("failed to stop sandbox(%q): %v", sandbox.ID, err)
+	}
+	// Update sandbox metadata
+	err = u.sandboxStore.Update(sandbox.ID, func(meta metadata.SandboxMetadata) (metadata.SandboxMetadata, error) {
+		meta.State = kubeapi.PodSandboxState_SANDBOX_NOTREADY
+		return meta, nil
+	})
+
+	return nil
+}
+
+// RemovePodSandbox deletes the sandbox. If there are any running containers in the
+// sandbox, they should be force deleted.
+func (u *UnikernelRuntime) RemovePodSandbox(podSandboxID string) error {
+	// Get sandbox and all containers in sandbox
+	sandbox, err := u.sandboxStore.Get(podSandboxID)
+	if err != nil {
+		if err == store.ErrNotExist {
```

Comment:
Is there any possible of leaking containers here?

Reply:
We always try to remove sandbox after removing all inside containers successfully, so I think it is fine here.